fireplace/cards/utils.py

```diff
@@ -38,6 +40,46 @@
 AT_MAX_MANA = lambda s: MANA(s) == 10
 
 
+def weighted_card_choice(source, weights: List[int], card_sets: List[str], count: int):
+	"""
+	take a list of weights and a list of card pools and produce a random weighted
+	sample without replacement.
+	len(weights) == len(card_sets) (one weight per card set)
+	"""
+
+	# special case for single weight (generic random pick)
+	if len(card_sets) == 1:
+		return [source.controller.card(card, source=source) for card in random.sample(card_sets[0], count)]
+
+	# special case for all weights equal
+	if len(set(weights)) == 1:
```

Comment:
I think the special case for a single weight is fine, but not this one.

In any case:

1. There's good reason to expect that both code paths have almost the same runtime
2. This is emphatically *not* in an inner loop of fireplace

If you really care about performance, profile or benchmark and then optimize where it will actually matters. Otherwise you'll end up wasting a lot of effort in optimizing things that *don't matter at all to actual performance*.